database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala

```diff
@@ -49,29 +54,38 @@ trait WorkflowStoreSlickDatabase extends WorkflowStoreSqlDatabase {
     runTransaction(action).void
   }
 
-  override def fetchStartableWorkflows(limit: Int)
+  private def now: Timestamp = new Timestamp(System.currentTimeMillis())
+
+  private def ago(duration: Duration): Timestamp = {
+    new Timestamp(System.currentTimeMillis() - duration.toMillis)
+  }
+
+  override def fetchStartableWorkflows(limit: Int, cromwellId: Option[String], heartbeatTtl: Duration)
                                       (implicit ec: ExecutionContext): Future[Seq[WorkflowStoreEntry]] = {
     val action = for {
-      workflowStoreEntries <- dataAccess.fetchStartableWorkflows(limit.toLong).result
-      _ <- DBIO.sequence(workflowStoreEntries map updateWorkflowStateAndRestartedForWorkflowExecutionUuid)
+      workflowStoreEntries <- dataAccess.fetchStartableWorkflows((limit.toLong, cromwellId, ago(heartbeatTtl))).result
+      _ <- DBIO.sequence(workflowStoreEntries map updateForFetched(cromwellId))
     } yield workflowStoreEntries
 
     runTransaction(action)
   }
 
-  private def updateWorkflowStateAndRestartedForWorkflowExecutionUuid(workflowStoreEntry: WorkflowStoreEntry)
-                                                         (implicit ec: ExecutionContext): DBIO[Unit] = {
+  private def updateForFetched(cromwellId: Option[String])(workflowStoreEntry: WorkflowStoreEntry)
+                                                          (implicit ec: ExecutionContext): DBIO[Unit] = {
     val workflowExecutionUuid = workflowStoreEntry.workflowExecutionUuid
     val updateState = workflowStoreEntry.workflowState match {
         // Submitted workflows become running when fetched
       case WorkflowStoreState.Submitted => WorkflowStoreState.Running
         // Running or Aborting stay as is
       case other => other
     }
+
+    val displayNow = DateFormat.getDateTimeInstance.format(now)
+
     for {
-      // When fetched, the restarted flag is set back to false so we don't pick it up next time.
-      updateCount <- dataAccess.workflowStateAndRestartedForWorkflowExecutionUuid(workflowExecutionUuid).update((updateState, false))
-      _ <- assertUpdateCount(s"Update $workflowExecutionUuid to $updateState", updateCount, 1)
+      // When fetched, the heartbeat timestamp is set to now so we don't pick it up next time.
+      updateCount <- dataAccess.workflowStateAndHeartbeatTimestampForWorkflowExecutionUuid(workflowExecutionUuid).update((updateState, Option(now)))
```

Comment:
shouldn't this set the `cromwellId` too ?

Reply:
yes it should. 😬